nhs/data/allocation.py

```diff
@@ -0,0 +1,286 @@
+    │ ---      ┆ ---         ┆ ---      ┆ ---     │
+    │ str      ┆ bool        ┆ f64      ┆ f64     │
+    ╞══════════╪═════════════╪══════════╪═════════╡
+    │ A        ┆ true        ┆ 1.0      ┆ 10.0    │
+    │ A        ┆ true        ┆ 1.0      ┆ 10.0    │
+    │ A        ┆ true        ┆ 2.0      ┆ 20.0    │
+    │ A        ┆ true        ┆ 3.0      ┆ 30.0    │
+    │ B        ┆ true        ┆ 4.0      ┆ 40.0    │
+    │ B        ┆ true        ┆ 4.0      ┆ 40.0    │
+    └──────────┴─────────────┴──────────┴─────────┘
+    """
+    if census.select(pl.len()).collect().item() == 0:
+        logger.warning("Empty census data provided.")
+        return census.select(pl.col(code_col, long_col, lat_col, feature_col))
+
+    return (
+        census.select(
+            pl.col(code_col, long_col, lat_col, feature_col)
+            # repeat (long, lat) to ensure num coordinates is bigger than sample size
+            .repeat_by(pl.col(feature_col)).flatten()
+        )
+        .filter(
+            pl.int_range(pl.len()).shuffle().over(code_col) < pl.col(feature_col)
+        )  # sample N rows
+        .with_columns(
+            pl.lit(True).alias(feature_col)
+        )  # fill the feature column with True
+    )
+
+
+@log_entry_exit()
+def _get_feat_non_feat_cols(feats_1: pl.LazyFrame, feats_2: pl.LazyFrame):
+    """
+    Get list of feature and non-feature columns from two LazyFrames.
+
+    **NOTE**: assumes `feats_1` and `feats_2` have the same schema that only
+    differs in the feature columns.
+    """
+    feats1_cols = feats_1.collect_schema().names()
+    feats2_cols = feats_2.collect_schema().names()
+    non_feat_cols = list(set(feats1_cols).intersection(feats2_cols))
+    feature_cols = list(set(feats1_cols).union(feats2_cols).difference(non_feat_cols))
+    return feature_cols, non_feat_cols
+
+
+@log_entry_exit()
+def _join_sampled_census_features(feats_1: pl.LazyFrame, feats_2: pl.LazyFrame):
+    """
+    Join two LazyFrames with sampled census features and coalesce the feature columns.
+    """
+    feat_cols, non_feat_cols = _get_feat_non_feat_cols(feats_1, feats_2)
+
+    return feats_1.join(
+        feats_2, on=non_feat_cols, how="full", coalesce=True
+    ).with_columns(pl.col(*feat_cols).fill_null(False))
+
+
+@log_entry_exit()
+def _stack_sampled_census_features(feats_1: pl.LazyFrame, feats_2: pl.LazyFrame):
+    """
+    Vertically stack two LazyFrames with sampled census features, uniting shared columns.
+    """
+    feat_cols, _ = _get_feat_non_feat_cols(feats_1, feats_2)
+
+    return pl.concat([feats_1, feats_2], how="diagonal_relaxed").with_columns(
+        pl.col(*feat_cols).fill_null(False)
+    )
+
+
+@log_entry_exit()
+def randomly_assign_census_features(
+    census: pl.LazyFrame,
+    code_col: str,
+    long_col: str,
+    lat_col: str,
+    feature_cols: list[str],
+    index_col: str = "person_id",
+    ignore_total: bool = True,
+    total_prefix: str = "Tot_",
+):
```

Comment:
This sampling approach will become redundant?

Reply:
Which one? You need to have *something* in order to make modifications so having it is not redundant either way even if we are changing it

Reply:
btw `_join_sampled...` is NOT used but I kept it as the join logic will be relevant when we try to make a person have multiple attributes